### PR TITLE
Adds xds requests metrics

### PR DESCRIPTION
**Pull request by pbetkier**
Adds measuring of gRPC xDS requests reported under `grpc.requests.<type>`, e.g. `grpc.requests.eds`.

Refactored and extended `MeteredConnectionsCallbacks` into `MetricsDiscoveryServerCallbacks` to implement this feature.

### CODE REVIEW

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/server/callbacks/MetricsDiscoveryServerCallbacks.kt [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    init {
+        connectionsByType = StreamType.values()
+            .map { type -> type to AtomicInteger(0) }
+            .toMap()
+
+        meterRegistry.gauge("grpc.all-connections", connections)
+        connectionsByType.forEach { (type, typeConnections) ->
+            meterRegistry.gauge("grpc.connections.${type.name.toLowerCase()}", typeConnections)
+        }
```
Comment (slonka) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/server/callbacks/MetricsDiscoveryServerCallbacks.kt:44]: As the diff does not work I'll just highlight the changes for other reviewers.

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/server/callbacks/MetricsDiscoveryServerCallbacks.kt [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        connectionsByType(typeUrl).incrementAndGet()
+    }
+
+    override fun onStreamClose(streamId: Long, typeUrl: String) {
+        connections.decrementAndGet()
+        connectionsByType(typeUrl).decrementAndGet()
+    }
+
+    override fun onV3StreamRequest(streamId: Long, request: V3DiscoveryRequest) {
+        meterRegistry.counter("grpc.requests.${StreamType.fromTypeUrl(request.typeUrl).name.toLowerCase()}")
```
Comment (slonka) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/server/callbacks/MetricsDiscoveryServerCallbacks.kt:58]: This was added.

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/server/callbacks/MetricsDiscoveryServerCallbacks.kt [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            .increment()
+    }
+
+    override fun onV2StreamRequest(streamId: Long, request: V2DiscoveryRequest) {
+        meterRegistry.counter("grpc.requests.${StreamType.fromTypeUrl(request.typeUrl).name.toLowerCase()}")
```
Comment (slonka) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/server/callbacks/MetricsDiscoveryServerCallbacks.kt:63]: This was added.

**Review by slonka — CHANGES_REQUESTED** [review]
You forgot about docs. https://envoy-control.readthedocs.io/en/latest/deployment/observability/#connections

File: envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/MetricsDiscoveryServerCallbacksTest.kt [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+            }
+        }
+    }
+
+    fun expectedGrpcConnectionsGaugeValues(): Map<MetricsDiscoveryServerCallbacks.StreamType, Int>
+
+    @Test
+    fun `should measure gRPC requests`() {
+        // given
+        val meterRegistry = envoyControl().app.bean(MeterRegistry::class.java)
```
Comment (pzmi) [inline envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/MetricsDiscoveryServerCallbacksTest.kt:132]: That looks so hacky

Comment (pzmi) [reply envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/MetricsDiscoveryServerCallbacksTest.kt:132]: Also, this is more of a question regarding `EnvoyControlExtension` and `EnvoyControlTestApp`. Is there anything wrong with more inline reify magic from `EnvoyControlTestConfiguration` so `EnvoyControlTestApp` does not do that?

**Review by pzmi — APPROVED** [review]

File: envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/MetricsDiscoveryServerCallbacksTest.kt (hunk 4 above)
Comment (pbetkier) [reply envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/MetricsDiscoveryServerCallbacksTest.kt:132]: What I can do is to hide meter registry extraction under `envoyControl().app.meterRegistry()` if you prefer :)

As for reify magic – that seemed more hacky and less explicit for me ;) maybe it's just my conservative Java background.

**Review by slonka — APPROVED** [review]

**Review by pzmi — APPROVED** [review]

**Pull request merged**